agent/controller/agent.go

```diff
@@ -98,82 +98,94 @@ func (a *Agent) Close(cause error) {
+func (a *Agent) processPacket(pkt *pb.Packet) {
+	sid := string(pkt.Spec[pb.SpecGatewaySessionID])
+	log.With("sid", sid).Debugf("received client packet [%v]", pkt.Type)
+	switch pkt.Type {
+	case pbagent.GatewayConnectOK:
+		log.Infof("connected with success to %v", a.config.URL)
+	case pbagent.SessionOpen:
+		a.processSessionOpen(pkt)
+
+	// terminal exec
+	case pbagent.ExecWriteStdin:
+		a.doExec(pkt)
+
+	case pbagent.SSMConnectionWrite:
+		a.processSSMProtocol(pkt)
+
+	// PG protocol
+	case pbagent.PGConnectionWrite:
+		a.processPGProtocol(pkt)
 
-		// terminal exec
-		case pbagent.ExecWriteStdin:
-			a.doExec(pkt)
+	// MySQL protocol
+	case pbagent.MySQLConnectionWrite:
+		a.processMySQLProtocol(pkt)
 
-		case pbagent.SSMConnectionWrite:
-			a.processSSMProtocol(pkt)
+	// MSSQL Protocol
+	case pbagent.MSSQLConnectionWrite:
+		a.processMSSQLProtocol(pkt)
 
-		// PG protocol
-		case pbagent.PGConnectionWrite:
-			a.processPGProtocol(pkt)
+	// MongoDB Protocol
+	case pbagent.MongoDBConnectionWrite:
+		a.processMongoDBProtocol(pkt)
 
-		// MySQL protocol
-		case pbagent.MySQLConnectionWrite:
-			a.processMySQLProtocol(pkt)
+	// raw tcp
+	case pbagent.TCPConnectionWrite:
+		a.processTCPWriteServer(pkt)
 
-		// MSSQL Protocol
-		case pbagent.MSSQLConnectionWrite:
-			a.processMSSQLProtocol(pkt)
+	// http proxy
+	case pbagent.HttpProxyConnectionWrite:
+		a.processHttpProxyWriteServer(pkt)
 
-		// MongoDB Protocol
-		case pbagent.MongoDBConnectionWrite:
-			a.processMongoDBProtocol(pkt)
+	// SSH protocol
+	case pbagent.SSHConnectionWrite:
+		a.processSSHProtocol(pkt)
 
-		// raw tcp
-		case pbagent.TCPConnectionWrite:
-			a.processTCPWriteServer(pkt)
+	// terminal
+	case pbagent.TerminalWriteStdin:
+		a.doTerminalWriteAgentStdin(pkt)
+	case pbagent.TerminalResizeTTY:
+		a.doTerminalResizeTTY(pkt)
 
-		// http proxy
-		case pbagent.HttpProxyConnectionWrite:
-			a.processHttpProxyWriteServer(pkt)
+	case pbagent.SessionClose:
+		a.processSessionClose(pkt)
 
-		// SSH protocol
-		case pbagent.SSHConnectionWrite:
-			a.processSSHProtocol(pkt)
+	case pbagent.TCPConnectionClose:
+		a.processTCPCloseConnection(pkt)
 
-		// terminal
-		case pbagent.TerminalWriteStdin:
-			a.doTerminalWriteAgentStdin(pkt)
-		case pbagent.TerminalResizeTTY:
-			a.doTerminalResizeTTY(pkt)
+	// system
+	case pbsystem.ProvisionDBRolesRequest:
+		dbprovisioner.ProcessDBProvisionerRequest(a.client, pkt)
 
-		case pbagent.SessionClose:
-			a.processSessionClose(pkt)
+	case pbsystem.RunbookHookRequestType:
+		runbookhook.ProcessRequest(a.client, pkt)
+	}
+}
 
-		case pbagent.TCPConnectionClose:
-			a.processTCPCloseConnection(pkt)
+func (a *Agent) Run() error {
+	a.client.StartKeepAlive()
 
-		// system
-		case pbsystem.ProvisionDBRolesRequest:
-			dbprovisioner.ProcessDBProvisionerRequest(a.client, pkt)
+	for {
+		pkt, err := a.client.Recv()
+		if err != nil {
+			if !errors.Is(err, io.EOF) {
+				log.Errorf("receiving channel error: %v", err)
+			}
+			break
```

Comment:
Just make sure to only return , the caller will handle of printing properly the context of the error